File: internal/features/sheepmatch/round.go
package sheepmatch

import (
	"context"
	"fmt"
	"strings"
	"time"

	"hdd/internal/client"
	"hdd/internal/logging"
	"hdd/internal/model"
	"hdd/internal/solver"
	"hdd/internal/terminal"
)

type actionPacer struct {
	minInterval time.Duration
	lastAction  time.Time
}

type roundProgress struct {
	current int
	total   int
}

func newActionPacer(minIntervalMs int) *actionPacer {
	return &actionPacer{minInterval: time.Duration(maxInt(0, minIntervalMs)) * time.Millisecond}
}

func (p *actionPacer) Wait() {
	if p == nil || p.minInterval <= 0 {
		return
	}
	if !p.lastAction.IsZero() {
		if wait := time.Until(p.lastAction.Add(p.minInterval)); wait > 0 {
			time.Sleep(wait)
		}
	}
	p.lastAction = time.Now()
}

func getConfigWithRetry(apiClient *client.APIClient, authState *runtimeAuthState) (*model.ConfigResponse, error) {
	var lastErr error
	for attempt := 0; attempt < 3; attempt++ {
		resp, err := withRuntimeAuthRetry(apiClient, authState, apiClient.GetConfig)
		if err == nil {
			return resp, nil
		}
		lastErr = err
		time.Sleep(200 * time.Millisecond)
	}
	return nil, lastErr
}

func getTileMeWithRetry(apiClient *client.APIClient, authState *runtimeAuthState) (*model.TileMeResponse, error) {
	return withRuntimeAuthRetry(apiClient, authState, apiClient.GetTileMe)
}

func getRemainingPlays(apiClient *client.APIClient, authState *runtimeAuthState, difficulty string) (int, error) {
	resp, err := getTileMeWithRetry(apiClient, authState)
	if err != nil {
		return 0, err
	}
	if resp == nil || resp.DailyPlaysRemaining == nil {
		return 0, nil
	}
	return resp.DailyPlaysRemaining[difficulty], nil
}

func nextRoundIndexForNewRound(usedToday int) int {
	if usedToday < 0 {
		usedToday = 0
	}
	return usedToday + 1
}

func totalRoundCount(usedToday int, remaining int) int {
	total := usedToday + remaining
	if total < 0 {
		return 0
	}
	return total
}

func normalizeRoundTotal(current int, total int) int {
	if current < 1 {
		current = 1
	}
	if total < current {
		return current
	}
	return total
}

func formatRoundProgress(current int, total int) string {
	total = normalizeRoundTotal(current, total)
	if current < 1 {
		current = 1
	}
	return fmt.Sprintf("今天第 %d/%d 局", current, total)
}

func formatUsedPowerups(usedPowerups []string) string {
	if len(usedPowerups) == 0 {
		return "没用道具"
	}
	labels := make([]string, 0, len(usedPowerups))
	for _, item := range usedPowerups {
		labels = append(labels, powerupLabel(item))
	}
	return "用了" + strings.Join(labels, "、")
}

func formatBalanceForDisplay(balance *float64) string {
	if balance == nil {
		return "未知"
	}
	return fmt.Sprintf("%.8f", *balance)
}

func printRoundResult(result model.RoundResultSummary) {
	total := normalizeRoundTotal(result.RoundIndex, result.RoundTotal)
	reason := ""
	if result.Err != nil {
		reason = fmt.Sprintf("，原因：%v", result.Err)
	}
	fmt.Printf("账号 %s 的%s难度%s结果：%s，%s，耗时 %s，奖励 %.8f，总余额 %s，今天还剩 %d 次，走了 %d 步%s。\n",
		result.Email,
		localizedDifficulty(result.Difficulty),
		formatRoundProgress(result.RoundIndex, total),
		roundStatusLabel(result),
		formatUsedPowerups(result.UsedPowerups),
		result.Duration.Round(time.Millisecond),
		result.Reward,
		formatBalanceForDisplay(result.BalanceAfter),
		result.RemainingAfter,
		result.MoveCount,
		reason,
	)
}

func drainPendingSessionsWithContext(ctx context.Context, apiClient *client.APIClient, authState *runtimeAuthState, dryRun bool, resultLogDir string, pacer *actionPacer, usedTodayByDifficulty map[string]int, remainingByDifficulty map[string]int) ([]model.RoundResultSummary, error) {
	if err := terminal.Check(ctx); err != nil {
		return nil, err
	}
	historyResp, err := withRuntimeAuthRetry(apiClient, authState, apiClient.GetHistory)
	if err != nil {
		return nil, err
	}
	var rounds []model.RoundResultSummary
	for _, item := range historyResp.Items {
		if err := terminal.Check(ctx); err != nil {
			return rounds, err
		}
		if item.Status != "pending" {
			continue
		}
		roundIndex := nextRoundIndexForNewRound(usedTodayByDifficulty[item.Difficulty])
		total := normalizeRoundTotal(roundIndex, totalRoundCount(usedTodayByDifficulty[item.Difficulty], remainingByDifficulty[item.Difficulty]))
		fmt.Printf("账号 %s 继续%s难度，%s（对局 %d）。\n", authState.account.Email, localizedDifficulty(item.Difficulty), formatRoundProgress(roundIndex, total), item.SessionID)
		start := historyItemToStartResponse(&item)
		round, err := playRoundWithContext(ctx, apiClient, authState, start, dryRun, resultLogDir, pacer, true, roundProgress{current: roundIndex, total: total})
		rounds = append(rounds, round)
		printRoundResult(round)
		if err != nil {
			return rounds, err
		}
	}
	return rounds, nil
}

func runDifficultyWithContext(ctx context.Context, apiClient *client.APIClient, authState *runtimeAuthState, difficulty string, dryRun bool, seed model.AccountRunSummary, resultLogDir string, pacer *actionPacer, nextRoundIndex int, totalRounds int) (model.AccountRunSummary, []model.RoundResultSummary) {
	summary := seed
	if err := terminal.Check(ctx); err != nil {
		summary.Err = err
		summary.When = time.Now()
		return summary, nil
	}
	if strings.TrimSpace(summary.Email) == "" {
		summary.Email = authState.account.Email
		summary.Difficulty = difficulty
	}
	remaining, err := getRemainingPlays(apiClient, authState, difficulty)
	if err != nil {
		summary.Err = err
		summary.When = time.Now()
		return summary, nil
	}
	if remaining <= 0 {
		summary.RemainingAfter = 0
		summary.When = time.Now()
		return summary, nil
	}
	totalRounds = normalizeRoundTotal(nextRoundIndex, totalRounds)
	var rounds []model.RoundResultSummary
	for played := 0; played < remaining; played++ {
		if err := terminal.Check(ctx); err != nil {
			summary.Err = err
			break
		}
		progress := roundProgress{current: nextRoundIndex + played, total: totalRounds}
		fmt.Printf("账号 %s 开始玩%s难度，%s。\n", authState.account.Email, localizedDifficulty(difficulty), formatRoundProgress(progress.current, progress.total))
		start, err := withRuntimeAuthRetry(apiClient, authState, func(authToken string) (*model.StartResponse, error) {
			return apiClient.StartGame(authToken, difficulty)
		})
		if err != nil {
			summary.Err = err
			break
		}
		round, roundErr := playRoundWithContext(ctx, apiClient, authState, start, dryRun, resultLogDir, pacer, false, progress)
		rounds = append(rounds, round)
		printRoundResult(round)
		summary = mergeRoundIntoSummary(summary, round)
		if roundErr != nil {
			summary.Err = roundErr
			break
		}
	}
	if len(rounds) == 0 {
		summary.RemainingAfter = remaining
		summary.When = time.Now()
	}
	if resultLogDir != "" {
		if err := logging.AppendDifficultySummary(resultLogDir, summary); err != nil {
			fmt.Printf("账号 %s 写入难度汇总失败：%v\n", authState.account.Email, err)
		}
	}
	return summary, rounds
}

func mergeRoundIntoSummary(summary model.AccountRunSummary, round model.RoundResultSummary) model.AccountRunSummary {
	if strings.TrimSpace(summary.Email) == "" {
		summary.Email = round.Email
		summary.Difficulty = round.Difficulty
	}
	summary.Played++
	summary.TotalReward += round.Reward
	summary.RemainingAfter = round.RemainingAfter
	summary.When = round.When
	if round.BalanceAfter != nil {
		summary.BalanceAfter = float64Ptr(*round.BalanceAfter)
	}
	if round.Err != nil {
		summary.Err = round.Err
		summary.Failed++
		return summary
	}
	switch round.Status {
	case "won":
		summary.Won++
	case "abandoned":
		summary.Abandoned++
	default:
		summary.Failed++
	}
	return summary
}

func playRoundWithContext(ctx context.Context, apiClient *client.APIClient, authState *runtimeAuthState, start *model.StartResponse, dryRun bool, resultLogDir string, pacer *actionPacer, continued bool, progress roundProgress) (model.RoundResultSummary, error) {
	snapshot := snapshotFromStartResponse(start)
	startedAt := time.Now()
	usedPowerups := make([]string, 0)
	for {
		if err := terminal.Check(ctx); err != nil {
			result := buildRoundError(authState.account.Email, snapshot, continued, dryRun, usedPowerups, startedAt, progress, err)
			if resultLogDir != "" {
				_ = logging.AppendRoundResult(resultLogDir, result)
			}
			return result, err
		}
		if snapshot.Status == "won" || (len(snapshot.Tiles) == 0 && len(snapshot.SlotTiles) == 0) {
			remaining, remainingErr := getRemainingPlays(apiClient, authState, snapshot.Difficulty)
			balance := extractBalanceFromStep(nil)
			result := model.RoundResultSummary{
				Email:          authState.account.Email,
				Difficulty:     snapshot.Difficulty,
				RoundIndex:     progress.current,
				RoundTotal:     progress.total,
				SessionID:      snapshot.SessionID,
				Continued:      continued,
				DryRun:         dryRun,
				Status:         "won",
				Reward:         0,
				BalanceAfter:   balance,
				RemainingAfter: remaining,
				MoveCount:      snapshot.MoveCount,
				UsedPowerups:   append([]string(nil), usedPowerups...),
				Duration:       time.Since(startedAt),
				When:           time.Now(),
				Err:            remainingErr,
			}
			if resultLogDir != "" {
				_ = logging.AppendRoundResult(resultLogDir, result)
			}
			return result, remainingErr
		}
		plan, err := solver.PlanToToolBoundary(snapshot)
		if err != nil {
			result := buildRoundError(authState.account.Email, snapshot, continued, dryRun, usedPowerups, startedAt, progress, err)
			if resultLogDir != "" {
				_ = logging.AppendRoundResult(resultLogDir, result)
			}
			return result, err
		}
		if len(plan.Actions) == 0 {
			result := buildRoundError(authState.account.Email, snapshot, continued, dryRun, usedPowerups, startedAt, progress, fmt.Errorf("当前整局无法生成可执行计划"))
			if resultLogDir != "" {
				_ = logging.AppendRoundResult(resultLogDir, result)
			}
			return result, result.Err
		}
		for _, action := range plan.Actions {
			if dryRun {
				var err error
				snapshot, err = applyPlannedActionLocally(snapshot, action)
				if err != nil {
					result := buildRoundError(authState.account.Email, snapshot, continued, dryRun, usedPowerups, startedAt, progress, err)
					if resultLogDir != "" {
						_ = logging.AppendRoundResult(resultLogDir, result)
					}
					return result, err
				}
				if action.Kind != "click" {
					usedPowerups = append(usedPowerups, action.Kind)
				}
				continue
			}
			if pacer != nil {
				pacer.Wait()
			}
			if err := terminal.Check(ctx); err != nil {
				result := buildRoundError(authState.account.Email, snapshot, continued, dryRun, usedPowerups, startedAt, progress, err)
				if resultLogDir != "" {
					_ = logging.AppendRoundResult(resultLogDir, result)
				}
				return result, err
			}
			stepResp, err := withRuntimeAuthRetry(apiClient, authState, func(authToken string) (*model.StepResponse, error) {
				return apiClient.Step(authToken, model.StepRequest{SessionID: snapshot.SessionID, Action: action.Kind, TileID: action.TileID})
			})
			if err != nil {
				if action.Kind == "click" && isSlotFullError(err) {
					next, localErr := applyLocalClickSnapshot(snapshot, action.TileID)
					if localErr == nil {
						snapshot = next
						continue
					}
				}
				result := buildRoundError(authState.account.Email, snapshot, continued, dryRun, usedPowerups, startedAt, progress, err)
				if resultLogDir != "" {
					_ = logging.AppendRoundResult(resultLogDir, result)
				}
				return result, err
			}
			snapshot = snapshotFromStepResponse(snapshot, stepResp)
			if action.Kind != "click" {
				usedPowerups = append(usedPowerups, action.Kind)
			}
			if stepResp.Status == "won" {
				remaining, remainingErr := getRemainingPlays(apiClient, authState, snapshot.Difficulty)
				result := model.RoundResultSummary{
					Email:          authState.account.Email,
					Difficulty:     snapshot.Difficulty,
					RoundIndex:     progress.current,
					RoundTotal:     progress.total,
					SessionID:      snapshot.SessionID,
					Continued:      continued,
					DryRun:         dryRun,
					Status:         stepResp.Status,
					Reward:         stepResp.RewardAmount,
					BalanceAfter:   float64Ptr(stepResp.Balance),
					RemainingAfter: remaining,
					MoveCount:      stepResp.MoveCount,
					UsedPowerups:   append([]string(nil), usedPowerups...),
					Duration:       time.Since(startedAt),
					When:           time.Now(),
					Err:            remainingErr,
				}
				if resultLogDir != "" {
					_ = logging.AppendRoundResult(resultLogDir, result)
				}
				return result, remainingErr
			}
		}
	}
}

func buildRoundError(email string, snapshot model.SessionSnapshot, continued bool, dryRun bool, usedPowerups []string, startedAt time.Time, progress roundProgress, err error) model.RoundResultSummary {
	return model.RoundResultSummary{
		Email:        email,
		Difficulty:   snapshot.Difficulty,
		RoundIndex:   progress.current,
		RoundTotal:   progress.total,
		SessionID:    snapshot.SessionID,
		Continued:    continued,
		DryRun:       dryRun,
		Status:       snapshot.Status,
		MoveCount:    snapshot.MoveCount,
		UsedPowerups: append([]string(nil), usedPowerups...),
		Duration:     time.Since(startedAt),
		When:         time.Now(),
		Err:          err,
	}
}

func extractBalanceFromStep(resp *model.StepResponse) *float64 {
	if resp == nil {
		return nil
	}
	return float64Ptr(resp.Balance)
}

func maxInt(left int, right int) int {
	if left > right {
		return left
	}
	return right
}

File: src/workflows/sheepmatch/round.rs
use std::collections::HashMap;
use std::io;
use std::sync::{Arc, Mutex};
use std::time::Instant;

use crate::api::ApiError;
use crate::model::{
    AccountRunSummary, ConfigResponse, RoundResultSummary, SessionSnapshot, StartResponse,
    StepRequest, StepResponse,
};
use crate::ui;
use crate::workflows::common::{
    current_unix_ms, humanize_retryable_api_error, is_pending_round_status, is_retryable_api_error,
};

use super::auth::{with_auth_retry, with_auth_retry_api};
use super::snapshot::{
    fixed_click_queue, is_slot_full_error, is_solved, is_stale_click_error,
    snapshot_from_start_response, snapshot_from_step_response,
};
use super::{AccountRuntime, BatchState};

const STEP_RETRY_BACKOFF: std::time::Duration = std::time::Duration::from_millis(500);
const STEP_RETRY_LOG_EVERY: usize = 10;

#[derive(Debug, Clone, Copy)]
pub(super) struct RoundProgress {
    pub(super) current: i32,
    pub(super) total: i32,
}

pub(super) fn play_round(
    cancel_flag: &ui::CancelFlag,
    state: &Arc<Mutex<BatchState>>,
    runtime: &mut AccountRuntime,
    config: &ConfigResponse,
    start: &StartResponse,
    continued: bool,
    progress: RoundProgress,
) -> io::Result<RoundResultSummary> {
    let started = Instant::now();
    let mut snapshot = snapshot_from_start_response(start);
    let used_powerups = Vec::new();
    let click_queue = fixed_click_queue(&snapshot);

    if click_queue.is_empty() {
        if is_solved(&snapshot) {
            let remaining_after =
                remaining_plays(state, runtime, &snapshot.difficulty).unwrap_or_default();
            return Ok(RoundResultSummary {
                email: runtime.email().to_string(),
                difficulty: snapshot.difficulty.clone(),
                round_index: progress.current,
                round_total: progress.total,
                session_id: snapshot.session_id,
                continued,
                dry_run: false,
                status: "won".to_string(),
                reward: 0.0,
                balance_after: None,
                remaining_after,
                move_count: snapshot.move_count,
                used_powerups,
                duration_ms: started.elapsed().as_millis() as i64,
                when_unix_ms: current_unix_ms(),
                error_message: String::new(),
            });
        }
        return Ok(build_round_error(
            runtime.email(),
            &snapshot,
            continued,
            progress,
            &used_powerups,
            started,
            "当前对局没有可点击的牌，无法继续。",
        ));
    }

    for tile_id in click_queue {
        ui::check_cancel(cancel_flag)?;
        if is_solved(&snapshot) {
            let remaining_after =
                remaining_plays(state, runtime, &snapshot.difficulty).unwrap_or_default();
            return Ok(RoundResultSummary {
                email: runtime.email().to_string(),
                difficulty: snapshot.difficulty.clone(),
                round_index: progress.current,
                round_total: progress.total,
                session_id: snapshot.session_id,
                continued,
                dry_run: false,
                status: "won".to_string(),
                reward: 0.0,
                balance_after: None,
                remaining_after,
                move_count: snapshot.move_count,
                used_powerups,
                duration_ms: started.elapsed().as_millis() as i64,
                when_unix_ms: current_unix_ms(),
                error_message: String::new(),
            });
        }
        if config.min_interval_ms > 0 {
            ui::sleep_with_cancel(
                cancel_flag,
                std::time::Duration::from_millis(config.min_interval_ms as u64),
            )?;
        }
        let mut step_attempts = 0usize;
        let response = loop {
            ui::check_cancel(cancel_flag)?;
            step_attempts += 1;
            let response = with_auth_retry_api(state, runtime, |client, auth_token| {
                client.step(
                    auth_token,
                    StepRequest {
                        session_id: snapshot.session_id,
                        action: "click".to_string(),
                        tile_id,
                    },
                )
            });
            match response {
                Err(ApiError::HttpStatus {
                    status: 409,
                    message,
                }) if is_slot_full_error(&message) || is_stale_click_error(&message) => {
                    break Err(io::Error::other(message));
                }
                Err(ApiError::HttpStatus { .. }) => {
                    ui::sleep_with_cancel(cancel_flag, STEP_RETRY_BACKOFF)?;
                    continue;
                }
                Err(error) if is_retryable_step_transport_error(&error) => {
                    if step_attempts == 1 || step_attempts.is_multiple_of(STEP_RETRY_LOG_EVERY) {
                        state.lock().unwrap().log.line_fmt(format_args!(
                            "账号 {} 的羊了个羊第 {} 步请求暂时失败，继续等待接口返回成功后再推进：{}",
                            runtime.email(),
                            snapshot.move_count + 1,
                            humanize_retryable_api_error(&error)
                        ));
                    }
                    ui::sleep_with_cancel(cancel_flag, STEP_RETRY_BACKOFF)?;
                    continue;
                }
                other => break other.map_err(|error| io::Error::other(error.to_string())),
            }
        };
        match response {
            Ok(step) => {
                let email = runtime.email().to_string();
                let remaining_after =
                    remaining_plays(state, runtime, &snapshot.difficulty).unwrap_or(0);
                let result = build_round_from_step(
                    &email,
                    &snapshot,
                    &step,
                    StepRoundContext {
                        continued,
                        progress,
                        used_powerups: &used_powerups,
                        started,
                        remaining_after,
                    },
                );
                snapshot = snapshot_from_step_response(&snapshot, &step);
                if is_terminal_round_status(&step.status) {
                    return Ok(result);
                }
            }
            Err(error) => {
                if is_slot_full_error(&error.to_string()) {
                    return Ok(build_round_error(
                        runtime.email(),
                        &snapshot,
                        continued,
                        progress,
                        &used_powerups,
                        started,
                        &error.to_string(),
                    ));
                }
                if is_stale_click_error(&error.to_string()) {
                    continue;
                }
                return Ok(build_round_error(
                    runtime.email(),
                    &snapshot,
                    continued,
                    progress,
                    &used_powerups,
                    started,
                    &error.to_string(),
                ));
            }
        }
    }

    if is_solved(&snapshot) {
        let remaining_after =
            remaining_plays(state, runtime, &snapshot.difficulty).unwrap_or_default();
        return Ok(RoundResultSummary {
            email: runtime.email().to_string(),
            difficulty: snapshot.difficulty.clone(),
            round_index: progress.current,
            round_total: progress.total,
            session_id: snapshot.session_id,
            continued,
            dry_run: false,
            status: snapshot.status.clone(),
            reward: 0.0,
            balance_after: None,
            remaining_after,
            move_count: snapshot.move_count,
            used_powerups,
            duration_ms: started.elapsed().as_millis() as i64,
            when_unix_ms: current_unix_ms(),
            error_message: String::new(),
        });
    }

    Ok(build_round_error(
        runtime.email(),
        &snapshot,
        continued,
        progress,
        &used_powerups,
        started,
        "这一局的可点击步骤已经用完，但服务端仍显示未通关。",
    ))
}

fn is_retryable_step_transport_error(error: &ApiError) -> bool {
    is_retryable_api_error(error)
}

fn is_terminal_round_status(status: &str) -> bool {
    matches!(
        status.trim().to_ascii_lowercase().as_str(),
        "won" | "lost" | "failed" | "game_over" | "abandoned"
    )
}

pub(super) fn next_round_index_for_new_round(used_today: i32) -> i32 {
    used_today.max(0) + 1
}

pub(super) fn total_round_count(used_today: i32, remaining: i32) -> i32 {
    (used_today + remaining).max(0)
}

pub(super) fn normalize_round_total(current: i32, total: i32) -> i32 {
    total.max(current.max(1))
}

pub(super) fn remaining_plays(
    state: &Arc<Mutex<BatchState>>,
    runtime: &mut AccountRuntime,
    difficulty: &str,
) -> io::Result<i32> {
    let me = with_auth_retry(state, runtime, |client, auth_token| {
        client.get_tile_me(auth_token)
    })?;
    Ok(*me.daily_plays_remaining.get(difficulty).unwrap_or(&0))
}

pub(super) fn summarize_rounds_by_difficulty(
    email: &str,
    rounds: &[RoundResultSummary],
) -> HashMap<String, AccountRunSummary> {
    let mut stats = HashMap::new();
    for round in rounds {
        let entry = stats
            .entry(round.difficulty.clone())
            .or_insert_with(|| AccountRunSummary {
                email: email.to_string(),
                difficulty: round.difficulty.clone(),
                ..AccountRunSummary::default()
            });
        merge_round_into_summary(entry, round);
    }
    stats
}

struct StepRoundContext<'a> {
    continued: bool,
    progress: RoundProgress,
    used_powerups: &'a [String],
    started: Instant,
    remaining_after: i32,
}

fn build_round_from_step(
    email: &str,
    snapshot: &SessionSnapshot,
    step: &StepResponse,
    context: StepRoundContext<'_>,
) -> RoundResultSummary {
    RoundResultSummary {
        email: email.to_string(),
        difficulty: snapshot.difficulty.clone(),
        round_index: context.progress.current,
        round_total: context.progress.total,
        session_id: snapshot.session_id,
        continued: context.continued,
        dry_run: false,
        status: step.status.clone(),
        reward: step.reward_amount,
        balance_after: Some(step.balance),
        remaining_after: context.remaining_after,
        move_count: step.move_count,
        used_powerups: context.used_powerups.to_vec(),
        duration_ms: context.started.elapsed().as_millis() as i64,
        when_unix_ms: current_unix_ms(),
        error_message: String::new(),
    }
}

fn build_round_error(
    email: &str,
    snapshot: &SessionSnapshot,
    continued: bool,
    progress: RoundProgress,
    used_powerups: &[String],
    started: Instant,
    error_message: &str,
) -> RoundResultSummary {
    RoundResultSummary {
        email: email.to_string(),
        difficulty: snapshot.difficulty.clone(),
        round_index: progress.current,
        round_total: progress.total,
        session_id: snapshot.session_id,
        continued,
        dry_run: false,
        status: snapshot.status.clone(),
        reward: 0.0,
        balance_after: None,
        remaining_after: 0,
        move_count: snapshot.move_count,
        used_powerups: used_powerups.to_vec(),
        duration_ms: started.elapsed().as_millis() as i64,
        when_unix_ms: current_unix_ms(),
        error_message: error_message.to_string(),
    }
}

pub(super) fn merge_round_into_summary(
    summary: &mut AccountRunSummary,
    result: &RoundResultSummary,
) {
    summary.played += 1;
    summary.total_reward += result.reward;
    summary.remaining_after = result.remaining_after;
    summary.when_unix_ms = result.when_unix_ms;
    if let Some(balance) = result.balance_after {
        summary.balance_after = Some(balance);
    }
    if !result.error_message.trim().is_empty() {
        summary.failed += 1;
        summary.error_message = result.error_message.clone();
        return;
    }
    let status = result.status.trim().to_ascii_lowercase();
    match status.as_str() {
        "won" => summary.won += 1,
        "abandoned" => summary.abandoned += 1,
        _ if is_pending_round_status(&status) => {}
        _ => summary.failed += 1,
    }
}
